tests/unit/lib/type_case_unittest.cc

```diff
@@ -0,0 +1,121 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#include <gtest/gtest.h>
+#include <shogun/lib/type_case.h>
+
+using namespace shogun;
+
+template <typename T>
+struct Simple
+{
+public:
+	Simple(T value_) : value(value_){};
+	T get_value()
+	{
+		return value;
+	};
+	void add_value(T value_)
+	{
+		value += value_;
+	};
+
+private:
+	T value;
+};
+
+TEST(Type_case, positional_lambdas)
+{
+	float32_t a_scalar = 42.0;
+	auto a_vector = SGVector<float32_t>({a_scalar});
+	auto a_matrix = SGMatrix<float32_t>(a_vector);
+	int counter = 0;
+
+	auto any_scalar = make_any(a_scalar);
+	auto any_vector = make_any(a_vector);
+	auto any_matrix = make_any(a_matrix);
+
+	auto f_scalar = [&counter](auto type) { counter++; };
+	auto f_vector = [&counter](auto type) { counter++; };
+	auto f_matrix = [&counter](auto type) { counter++; };
+
+	sg_for_each_type(any_scalar, sg_all_types, f_scalar);
+	EXPECT_EQ(counter, 1);
+
+	sg_for_each_type(any_vector, sg_all_types, nullptr, f_vector);
+	EXPECT_EQ(counter, 2);
+
+	sg_for_each_type(any_matrix, sg_all_types, nullptr, nullptr, f_matrix);
```

Comment:
this API is not the most elegant....need to discuss this

Reply:
Hmmm ok! Can have three different functions, it could make my life easier when specialising the helper function templates

Reply:
I wonder: Can't this be implicitly be done by forcing the lambda to have a different signature (including SGVector/matrix) ....
If not, I think having three functions is the best bet, at least API wise. They all could call the function you wrote here internally to avoid redundant code

Reply:
I think that could be possible, assuming that the compiler can implicitly convert SGVector and SGMatrix specialisations, right? Like in the summaryValue->set_simple_value(float value), I think all scalars can be interpreted as float right?

Btw seems like `std::is_scalar` is quite a broad definition, will have to find a more strict definition like `std::is_arithmetic` (potentially combined with a `std::decay`)

Reply:
> I think that could be possible, assuming that the compiler can implicitly convert SGVector and SGMatrix specialisations, right?
What do you mean by that? BTW not sure if that helps, but It is definitely possible to extract inner types from templates, see [this ugly thing](https://github.com/shogun-toolbox/shogun/blob/develop/src/interfaces/swig/shogun.i#L151) for example

> Like in the summaryValue->set_simple_value(float value), I think all scalars can be interpreted as float right?
That is true, but the compiler will not implicitly convert `float` to `SGVector`
> 

> Btw seems like `std::is_scalar` is quite a broad definition, will have to find a more strict definition like `std::is_arithmetic` (potentially combined with a `std::decay`). 
Yes definitely, I would have a typedef for that that checks, see [here](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/base/base_types.h)